src/Microsoft.ML.Transforms/TensorflowTransform.cs

```diff
@@ -0,0 +1,501 @@
+            {
+                var tfoutput = new TFOutput(tfSession.Graph[columnName]);
+                var shape = tfSession.Graph.GetTensorShape(tfoutput);
+
+                int[] dims = new int[shape.NumDimensions];
+                for (int k = 0; k < shape.NumDimensions; k++)
+                {
+                    dims[k] = (int)(shape[k] == -1 ? batchSize : shape[k]);
+                }
+                // Output tensor is expected to be of shape [Batch, Data]
+               return new VectorType(PrimitiveType.FromKind(DataKind.R4), dims);
+            }
+
+            private TFColInfo BuildTFInputMetaData(TFSession tfSession, int batchSize, string[] source)
+            {
+                var tfShapes = new TFShape[source.Length];
+                var tfTypes = new TFDataType[source.Length];
+                var colNames = new string[source.Length];
+                for (int j = 0; j < source.Length; j++)
+                {
+                    colNames[j] = source[j];
+                    var tfoutput = new TFOutput(tfSession.Graph[colNames[j]]);
+                    tfShapes[j] = tfSession.Graph.GetTensorShape(tfoutput);
+                    tfTypes[j] = tfoutput.OutputType;
+
+                    var l = new long[tfShapes[j].NumDimensions];
+                    for (int ishape = 0; ishape < tfShapes[j].NumDimensions; ishape++)
+                    {
+                        l[ishape] = tfShapes[j][ishape] == -1 ? batchSize : tfShapes[j][ishape];
+                    }
+                    tfShapes[j] = new TFShape(l);
+                }
+                return new TFColInfo(colNames, tfShapes, tfTypes);
+            }
+
+            protected override ColumnType GetColumnTypeCore(int iinfo)
+            {
+                Contracts.Assert(0 <= iinfo & iinfo < OutputCols.Length);
+
+                Contracts.Assert(OutputCols[iinfo] != null);
+                return OutputCols[iinfo];
+            }
+
+            private static string TestTypes(ColumnType[] types)
+            {
+                return null;
+            }
+        }
+
+        public const string Summary = "Transforms the data using the tenorflow model.";
+        public const string UserName = "TensorflowTransform";
+        public const string ShortName = "TFTransform";
+
+        public const string LoaderSignature = "TFTransform";
+        private const string RegistrationName = "Tensorflow";
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                modelSignature: "TENSFLOW",
+                verWrittenCur: 0x00010001, // Initial
+                verReadableCur: 0x00010001,
+                verWeCanReadBack: 0x00010001,
+                loaderSignature: LoaderSignature);
+        }
+
+        private readonly Bindings _bindings;
+
+        /// <summary>
+        /// Tensorflow session object
+        /// </summary>
+        private TFSession _session;
```

Comment:
>_session [](start = 26, length = 8)

Need to find out a way to dispose this safely.

Reply:
Instead of implementing IDisposable, you can dispose of this resource in the finalizer of the transform.

---
In reply to: [210042898](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210042898) [](ancestors = 210042898)